### PR TITLE
Add yum repo for RPM of OS and OSD in staging

### LINKED ISSUE #1549: [RPM M4] Publish YUM repository
Tasks | Estimate | Status | Notes | Dependencies
-- | -- | -- | -- | --
Deploy the RPM packages into a local YUM repository on a build system | 3 | Complete | The RPM package needs to be ready | see https://github.com/opensearch-project/opensearch-build/issues/1549#issuecomment-1076914878
The metadata files for the entire repo will be generated and staged in the repo | 2 | Complete | Need to figure out whether createrepo is the best way to do this | same as above
The build system should try to install the RPM packages through the ~~local~~ remote repo and verify the commands running correctly | 2 | Complete (We moved the local yum repo to remote staging yum repo) | The YUM repo needs to be ready | same as above
The build system should attempt to start the service of the installed packages and verify the outcome through process manager of the service (see more details in the test requirements) | 2 | Please track in #1555  | | see https://github.com/opensearch-project/opensearch-build/issues/1549#issuecomment-1076918348
~~Once verified,~~ the entire repo should be uploaded to the dedicated production S3 bucket linked to artfacts.opensearch.org | 2 | Complete |   | 
A repo file is also generated (once every major release) and updated on our website, for user to download and deploy on their host / server | 1 | Complete |   |  
Once deployed, user can then use YUM command to install the latest RPM packages from our repository on S3, with the help of CloudFront cache in the front | 2 | Complete |   |  

* We will attempt to have one repo per major version so that:
  * If users want to only install 1.x, they install 1.x yum repo file, and yum update will only bring them latest version in 1.x.
  * If users want to upgrade to 2.x, they install 2.x yum repo file, and since they already installed 1.x repo file, they can easily yum upgrade to 2.x, then yum downgrade back to their original 1.x version.
  * If users want to install 2.x directly without any old versions on their host, they can install 2.x yum repo file and install the latest version in 2.x.

* 20220414: The above chart is for staging yum repo which we have already implemented. Production repo will be made manually for now and will implement automation soon. See x64 examples:
  * https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/2.0.0-rc1/latest/linux/x64/rpm/dist/opensearch/opensearch-2.0.0-rc1-staging.repo
  * https://ci.opensearch.org/ci/dbc/distribution-build-opensearch-dashboards/2.0.0-rc1/latest/linux/x64/rpm/dist/opensearch-dashboards/opensearch-dashboards-2.0.0-rc1-staging.repo

### PR DESCRIPTION
**Pull request by tianleh**
### Description
Add the logic to create yum repo data and file for RPM of both OpenSearch and OpenSearch Dashboards.
 
### Issues Resolved
#1549
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by tianleh** [comment]
Tested in my test Jenkins job http://jenki-jenki-fpgmrv2ryxko-1366042710.us-east-1.elb.amazonaws.com/job/Playground/job/tianleh-test/job/tianleh-os-yum-4-12/ (see build number 12)

https://ci.opensearch.org/ci/dbc/Playground/tianleh-test/tianleh-os-yum-4-12/3.0.0/12/linux/x64/rpm/dist/opensearch/opensearch-artifacts.repo

@peterzhuamazon Please help double check the content of the repo file and data in S3. (The path can be found based on the Jenkins link above)

File: vars/buildYumRepo.groovy [hunk 1]
```diff
@@ -0,0 +1,27 @@
+void call(Map args = [:]) {
+    lib = library(identifier: 'jenkins@20211123', retriever: legacySCM(scm))
+
+    def buildManifest = lib.jenkins.BuildManifest.new(readYaml(file: args.buildManifest))
+    echo "Creating repo file and data for ${args.buildManifest}"
```
Comment (peterzhuamazon) [inline vars/buildYumRepo.groovy:5]: `Create yum repo metadata and repo file for ......`

File: vars/buildYumRepo.groovy [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    def repoFileContent = [
+        "[${filename}-staging-${version}-${BUILD_NUMBER}]",
+        "name=${name} ${version} ${BUILD_NUMBER} Staging",
+        "baseurl=${args.baseUrl}/rpm/dist/${filename}/",
+        "enabled=1",
+        "autorefresh=1",
+        "type=rpm-md"
+    ].join('\n')
+
+    writeFile file: "${repoFilePath}/${filename}-artifacts.repo", text: repoFileContent
```
Comment (peterzhuamazon) [inline vars/buildYumRepo.groovy:26]: The file name better match the content:
```
         ${filename}-staging-${version}.repo
```

File: vars/buildYumRepo.groovy (hunk 1 above)
Comment (tianleh) [reply vars/buildYumRepo.groovy]: updated

File: vars/buildYumRepo.groovy (hunk 2 above)
Comment (tianleh) [reply vars/buildYumRepo.groovy]: updated

Comment (peterzhuamazon) [reply vars/buildYumRepo.groovy]: Actually it is better to be `${filename}-${version}.staging.repo`
The production can be `${filename}-${version}.repo`

Comment (tianleh) [reply vars/buildYumRepo.groovy]: sure

**Comment by peterzhuamazon** [comment]
Thanks @tianleh please add some test cases.
Thanks.

**Comment by tianleh** [comment]
new repo file name https://ci.opensearch.org/ci/dbc/Playground/tianleh-test/tianleh-os-yum-4-12/3.0.0/14/linux/x64/rpm/dist/opensearch/opensearch-3.0.0.staging.repo

**Comment by tianleh** [comment]
> Thanks @tianleh please add some test cases.
> Thanks.

Sure. will add the unit tests before marking the PR ready for review.

**Comment by tianleh** [comment]
added the repo file to `/etc/yum.repos.d` and then install

```

 yum install opensearch.x86_64
Loaded plugins: fastestmirror, ovl
Loading mirror speeds from cached hostfile
 * base: download.cf.centos.org
 * epel: d2lzkl7pfhq30w.cloudfront.net
 * extras: download.cf.centos.org
 * updates: download.cf.centos.org
Resolving Dependencies
--> Running transaction check
---> Package opensearch.x86_64 0:3.0.0-1 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

==============================================================================================================================================================================
 Package                                Arch                               Version                              Repository                                               Size
==============================================================================================================================================================================
Installing:
 opensearch                             x86_64                             3.0.0-1                              opensearch-staging-3.0.0-14                             187 M

Transaction Summary
==============================================================================================================================================================================
Install  1 Package

Total download size: 187 M
Installed size: 384 M
Is this ok [y/d/N]: y
Downloading packages:
Package opensearch-3.0.0-linux-x64.rpm is not signed                     99% [=============================================================-]  20 MB/s | 186 MB  00:00:00 ETA 
opensearch-3.0.0-linux-x64.rpm                                                                                                                         | 187 MB  00:00:09     


Package opensearch-3.0.0-linux-x64.rpm is not signed
```

**Comment by tianleh** [comment]
during osd 2.0 core testing, notice this rpm build error

```
2022-04-13 19:38:16 INFO     Execute rpmbuild -bb --define '_topdir /tmp/tmp3aa32vov' --define '_version 2.0.0-rc1' --define '_architecture x86_64' opensearch-dashboards.rpm.spec in /tmp/tmp3aa32vov
error: line 17: Illegal char '-' in: Version: 2.0.0-rc1
Traceback (most recent call last):
  File "./src/run_assemble.py", line 58, in <module>
    sys.exit(main())
  File "./src/run_assemble.py", line 49, in main
    bundle.package(output_dir)
  File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/bundle.py", line 102, in package
    self.min_dist.build(self.bundle_recorder.package_name, dest)
  File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/dist.py", line 82, in build
    self.__build__(name, dest)
  File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/dist.py", line 118, in __build__
    BundleRpm(self.filename, self.path, self.min_path).build(name, dest, self.archive_path, self.build_cls)
  File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/bundle_rpm.py", line 111, in build
    subprocess.check_call(bundle_cmd, cwd=ext_dest, shell=True)
  File "/usr/local/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'rpmbuild -bb --define '_topdir /tmp/tmp3aa32vov' --define '_version 2.0.0-rc1' --define '_architecture x86_64' opensearch-dashboards.rpm.spec' returned non-zero exit status 1.
```

**Comment by dblock** [comment]
> during osd 2.0 core testing, notice this rpm build error
> 
> ```
> 2022-04-13 19:38:16 INFO     Execute rpmbuild -bb --define '_topdir /tmp/tmp3aa32vov' --define '_version 2.0.0-rc1' --define '_architecture x86_64' opensearch-dashboards.rpm.spec in /tmp/tmp3aa32vov
> error: line 17: Illegal char '-' in: Version: 2.0.0-rc1
> Traceback (most recent call last):
>   File "./src/run_assemble.py", line 58, in <module>
>     sys.exit(main())
>   File "./src/run_assemble.py", line 49, in main
>     bundle.package(output_dir)
>   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/bundle.py", line 102, in package
>     self.min_dist.build(self.bundle_recorder.package_name, dest)
>   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/dist.py", line 82, in build
>     self.__build__(name, dest)
>   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/dist.py", line 118, in __build__
>     BundleRpm(self.filename, self.path, self.min_path).build(name, dest, self.archive_path, self.build_cls)
>   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/bundle_rpm.py", line 111, in build
>     subprocess.check_call(bundle_cmd, cwd=ext_dest, shell=True)
>   File "/usr/local/lib/python3.7/subprocess.py", line 363, in check_call
>     raise CalledProcessError(retcode, cmd)
> subprocess.CalledProcessError: Command 'rpmbuild -bb --define '_topdir /tmp/tmp3aa32vov' --define '_version 2.0.0-rc1' --define '_architecture x86_64' opensearch-dashboards.rpm.spec' returned non-zero exit status 1.
> ```

So the tool, rpmbuild doesn't agree with semver? https://github.com/semver/semver/issues/145

It seems like [the recommendation](https://github.com/semver/semver/issues/145#issuecomment-851026311) is to change the version for the rpm package, i.e. replace `-alpha1` by `.alpha1`. I think it's ok. It won't matter for the final release anyway. WDYT?

**Comment by peterzhuamazon** [comment]
> > during osd 2.0 core testing, notice this rpm build error
> > ```
> > 2022-04-13 19:38:16 INFO     Execute rpmbuild -bb --define '_topdir /tmp/tmp3aa32vov' --define '_version 2.0.0-rc1' --define '_architecture x86_64' opensearch-dashboards.rpm.spec in /tmp/tmp3aa32vov
> > error: line 17: Illegal char '-' in: Version: 2.0.0-rc1
> > Traceback (most recent call last):
> >   File "./src/run_assemble.py", line 58, in <module>
> >     sys.exit(main())
> >   File "./src/run_assemble.py", line 49, in main
> >     bundle.package(output_dir)
> >   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/bundle.py", line 102, in package
> >     self.min_dist.build(self.bundle_recorder.package_name, dest)
> >   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/dist.py", line 82, in build
> >     self.__build__(name, dest)
> >   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/dist.py", line 118, in __build__
> >     BundleRpm(self.filename, self.path, self.min_path).build(name, dest, self.archive_path, self.build_cls)
> >   File "/var/jenkins/workspace/Playground/tianleh-test/tianleh-osd-yum-4-12/src/assemble_workflow/bundle_rpm.py", line 111, in build
> >     subprocess.check_call(bundle_cmd, cwd=ext_dest, shell=True)
> >   File "/usr/local/lib/python3.7/subprocess.py", line 363, in check_call
> >     raise CalledProcessError(retcode, cmd)
> > subprocess.CalledProcessError: Command 'rpmbuild -bb --define '_topdir /tmp/tmp3aa32vov' --define '_version 2.0.0-rc1' --define '_architecture x86_64' opensearch-dashboards.rpm.spec' returned non-zero exit status 1.
> > ```
> 
> So the tool, rpmbuild doesn't agree with semver? [semver/semver#145](https://github.com/semver/semver/issues/145)
> 
> It seems like [the recommendation](https://github.com/semver/semver/issues/145#issuecomment-851026311) is to change the version for the rpm package, i.e. replace `-alpha1` by `.alpha1`. I think it's ok. It won't matter for the final release anyway. WDYT?

Transfer discussion to https://github.com/opensearch-project/opensearch-build/issues/1549#issuecomment-1098429590

**Comment by peterzhuamazon** [comment]
@tianleh something seems not right as the manifest version number should not contain `rc1`.

**Comment by tianleh** [comment]
successful osd 2.0 build (without plugins) http://jenki-jenki-fpgmrv2ryxko-1366042710.us-east-1.elb.amazonaws.com/job/Playground/job/tianleh-test/job/tianleh-osd-yum-4-12/9/

repo file https://ci.opensearch.org/ci/dbc/Playground/tianleh-test/tianleh-osd-yum-4-12/2.0.0-rc1/9/linux/x64/rpm/dist/opensearch-dashboards/opensearch-dashboards-2.0.0-rc1.staging.repo

**Review by seraphjiang — APPROVED** [review]

**Comment by tianleh** [comment]
> @tianleh something seems not right as the manifest version number should not contain `rc1`.

Looks like not related to this change. It is more a general issue of `rc1` support? @peterzhuamazon

**Comment by peterzhuamazon** [comment]
> > @tianleh something seems not right as the manifest version number should not contain `rc1`.
> 
> Looks like not related to this change. It is more a general issue of `rc1` support? @peterzhuamazon

Ignore this as I have created issue on #1965

**Comment by dblock** [comment]
> > > @tianleh something seems not right as the manifest version number should not contain `rc1`.
> > 
> > 
> > Looks like not related to this change. It is more a general issue of `rc1` support? @peterzhuamazon
> 
> Ignore this as I have created issue on #1965

FYI I am not working on it

**Review by peterzhuamazon — APPROVED** [review]

**Pull request merged**